src/Features/Core/Portable/UnusedReferences/UnusedReferencesService.cs

```diff
@@ -0,0 +1,199 @@
+            }
+
+            HashSet<string> usedAssemblyLookup = new(compilation.GetUsedAssemblyReferences(cancellationToken)
+                .Select(reference => reference.Display)
+                .OfType<string>());
+
+            return GetUnusedReferences(usedAssemblyLookup, references);
+        }
+
+        internal static ImmutableArray<ReferenceInfo> GetUnusedReferences(
+            HashSet<string> usedAssemblyLookup,
+            ImmutableArray<ReferenceInfo> references)
+        {
+            var unusedReferencesBuilder = ImmutableArray.CreateBuilder<ReferenceInfo>();
+            var referencesByType = references.GroupBy(reference => reference.ReferenceType)
+                .ToImmutableDictionary(group => group.Key);
+
+            // We process the references in order by their type. This means we favor transitive references
+            // over direct references where possible.
+            foreach (var referenceType in _processingOrder)
+            {
+                if (!referencesByType.TryGetValue(referenceType, out var referencesForReferenceType))
+                {
+                    continue;
+                }
+
+                AddUnusedReferences(
+                    referencesForReferenceType.ToImmutableArray(),
+                    usedAssemblyLookup,
+                    unusedReferencesBuilder);
+            }
+
+            return unusedReferencesBuilder.ToImmutableArray();
+        }
+
+        private static void AddUnusedReferences(
+            ImmutableArray<ReferenceInfo> references,
+            HashSet<string> usedAssemblyLookup,
+            ImmutableArray<ReferenceInfo>.Builder unusedReferencesBuilder)
+        {
+
+            // This method checks for used reference two different ways.
+
+            // #1. We check if a reference directly brings in a used compilation assembly.
+            //
+            //    references: [ PackageReference(compilationAssembly: "/libs/Used.dll") ],
+            //    usedAssemblyLookup: [ "/libs/Used.dll" ]
+            //
+
+            // #2. We check if a reference transitively brings in a used compilation assembly.
+            //
+            //    references: [
+            //      ProjectReference(
+            //        compilationAssembly: "/libs/Unused.dll",
+            //        dependencies: [ PackageReference(compilationAssembly: "/libs/Used.dll") ]
+            //      ) ]
+            //    usedAssemblyLookup: [ "/libs/Used.dll" ]
+
+            // Check #1. we will look at the compilation assemblies brought in directly by the
+            // references to see if they are used.
+            var usedDirectReferences = references.Where(reference
+                => reference.CompilationAssemblies.Any(usedAssemblyLookup.Contains)).ToArray();
+
+            // We then want to gather all the assemblies brought in directly or transitively by
+            // these used assemblies so that we can remove them from our lookup.
```

Comment:
Why are we removing them?

Reply:
The way I approached this problem is that we have a list of used assemblies and a list of references. Our job is to to determine which references bring in used assemblies and which don't.  Once we know a reference is "used", meaning brings in a used assembly, then we have answered the question of which reference is responsible for bringing in those assemblies. So we remove them from our lookup and proceed to determine the source of the remaining used assemblies. Remaining references will need to bring in a different used assembly into the compilation to be considered necessary.